hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/files/StreamFileProducerBase.java

```diff
@@ -0,0 +1,355 @@
+        // pick record file format
+        recordFileFormat = switch (recordFileVersion) {
+            case 6 -> RecordFileFormatV6.INSTANCE;
+            default -> throw new IllegalArgumentException("Unknown record file version: " + recordFileVersion);};
+    }
+
+    // =========================================================================================================================================================================
+    // public abstract methods
+
+    /**
+     * Set the current running hash of record stream items. This is called only once to initialize the running hash on startup.
+     * This is called on handle transaction thread.
+     *
+     * @param recordStreamItemRunningHash The new running hash, all future running hashes produced will flow from this
+     */
+    public abstract void setRunningHash(Bytes recordStreamItemRunningHash);
+
+    /**
+     * Get the current running hash of record stream items. This is called on handle transaction thread. It will block
+     * if background thread is still hashing. It will always return the running hash after the last user transaction
+     * was added. Hence, any pre-transactions or others not yet committed via
+     * {@link StreamFileProducerBase#writeRecordStreamItems(long, Instant, Stream)} will not be included.
+     *
+     * @return The current running hash upto and including the last record stream item sent in writeRecordStreamItems().
+     */
+    @Nullable
+    public abstract Bytes getRunningHash();
+
+    /**
+     * Get the previous, previous, previous runningHash of all RecordStreamObject. This will block if
+     * the running hash has not yet been computed for the most recent user transaction.
+     *
+     * @return the previous, previous, previous runningHash of all RecordStreamObject
+     */
+    @Nullable
+    public abstract Bytes getNMinus3RunningHash();
+
+    /**
+     * Called at the end of a block and start of next block.
+     * <p>If there is a currently open block it closes any open stream files and writes signature file.</p>
+     * <p>Then opens a new record file for the new block</p>
+     *
+     * @param lastBlockNumber                       The number for the block we are closing. If lastBlockNumber is &lt;=0 then
+     *                                              there is no block to close.
+     * @param newBlockNumber                        The number for the block we are opening.
+     * @param newBlockFirstTransactionConsensusTime The consensus time of the first transaction in the new block. It must be the
+     *                                              adjusted consensus time not the platform assigned consensus time. Assuming
+     *                                              the two are different.
+     */
+    public abstract void switchBlocks(
+            long lastBlockNumber, long newBlockNumber, Instant newBlockFirstTransactionConsensusTime);
```

Comment:
Do we need both `lastBlockNumber` and `newBlockNumber`? Is it always true that `newBlockNumber` is `lastBlockNumber + 1`?